File: sdk-python/ext_echo.py
#!/usr/bin/env python3
"""Reference echo on the Python SDK (M2.4).

Usage: ``ext_echo.py --matrix-sock <sock> --id <logical>``
Same behavior as the Rust echo: abortable sleep, remote error, opening
count, cancel mark, stream flood, and clean exit.
"""

import argparse
import sys
import threading
import time

sys.path.insert(0, __file__.rsplit("/", 1)[0])

from matrix_component import CallCtx, Component, Handler


def append_line(path: str, line: str) -> None:
    try:
        with open(path, "a", encoding="utf-8") as f:
            f.write(line + "\n")
    except OSError:
        pass


class Echo(Handler):
    def __init__(self):
        self.marks: dict[str, str] = {}
        self.lock = threading.Lock()

    def on_call(self, ctx: CallCtx, ticket: str, cap: str, input: dict, cancel: threading.Event):
        if not isinstance(input, dict):
            input = {}
        cf = input.get("count_file")
        if isinstance(cf, str):
            append_line(cf, ticket)
        mc = input.get("mark_cancel")
        if isinstance(mc, str):
            with self.lock:
                self.marks[ticket] = mc
        sleep_ms = input.get("sleep_ms", 0)
        try:
            sleep_ms = int(sleep_ms)
        except (TypeError, ValueError):
            sleep_ms = 0
        slept = 0
        while slept < sleep_ms:
            if cancel.is_set():
                return ("error", "cancelled", "aborted")
            time.sleep(0.005)
            slept += 5
        fail = input.get("fail")
        if isinstance(fail, str):
            with self.lock:
                self.marks.pop(ticket, None)
            return ("error", fail, f"remote {fail}")
        fl = input.get("flood_stream")
        if isinstance(fl, dict):
            fid = fl.get("id", "s1")
            try:
                chunk = min(int(fl.get("chunk", 65536)), 1 << 20)
            except (TypeError, ValueError):
                chunk = 65536
            try:
                count = min(int(fl.get("count", 4)), 64)
            except (TypeError, ValueError):
                count = 4
            payload = "x" * chunk
            for seq in range(count):
                if cancel.is_set():
                    return ("error", "cancelled", "aborted")
                try:
                    ctx.send_stream(fid, seq, payload)
                except OSError:
                    break
        with self.lock:
            self.marks.pop(ticket, None)
        return {"echo": input}

    def on_cancel(self, ticket: str):
        with self.lock:
            path = self.marks.pop(ticket, None)
        if path is not None:
            append_line(path, ticket)


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--matrix-sock", required=True)
    ap.add_argument("--id", default="ext-echo")
    args = ap.parse_args()
    try:
        comp = Component.connect(args.matrix_sock, args.id)
    except (OSError, AssertionError, ValueError) as exc:
        print(f"connect: {exc}", file=sys.stderr)
        return 2
    reason = comp.serve(Echo())
    return 0 if reason in ("dispose", "eof") else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/matrix-host/examples/ext_echo.rs
//! Reference echo over the Rust SDK (M2.4).
//!
//! Usage: `ext_echo --matrix-sock <sock> --id <logical>`
//! Same behavior as the manual example it replaces:
//! - `input.sleep_ms` sleeps abortably (5ms);
//! - `input.fail = "CODE"` replies with a remote business error;
//! - `input.count_file = path`: appends the ticket per received call;
//! - `input.mark_cancel = path`: appends the ticket on observing `call.cancel`;
//! - `input.flood_stream = {id, chunk, count}`: sends N `stream.data`;
//! - otherwise replies `output = {"echo": input}`;
//! - `lifecycle.dispose` replies and exits 0 (via `serve`).

use matrix_component::{CallCtx, CallOutcome, Component, Handler};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::{
    atomic::{AtomicBool, Ordering},
    Mutex,
};
use std::time::Duration;

struct Echo {
    /// Tickets with pending cancel marks (consumed in `on_cancel`).
    marks: Mutex<HashMap<String, String>>,
}

fn append_line(path: &str, line: &str) {
    use std::io::Write;
    if let Ok(mut f) = std::fs::OpenOptions::new().create(true).append(true).open(path) {
        let _ = writeln!(f, "{}", line);
    }
}

fn abortable_sleep(ms: u64, cancel: &AtomicBool) -> bool {
    let mut slept = 0u64;
    while slept < ms {
        if cancel.load(Ordering::SeqCst) {
            return true;
        }
        std::thread::sleep(Duration::from_millis(5));
        slept += 5;
    }
    false
}

impl Handler for Echo {
    fn on_call(
        &self,
        ctx: &CallCtx,
        ticket: &str,
        _cap: &str,
        input: &Value,
        cancel: &AtomicBool,
    ) -> CallOutcome {
        if let Some(cf) = input.get("count_file").and_then(|v| v.as_str()) {
            append_line(cf, ticket);
        }
        if let Some(mc) = input.get("mark_cancel").and_then(|v| v.as_str()) {
            self.marks.lock().unwrap().insert(ticket.to_string(), mc.to_string());
        }
        let sleep_ms = input.get("sleep_ms").and_then(|v| v.as_u64()).unwrap_or(0);
        if abortable_sleep(sleep_ms, cancel) {
            // `on_cancel` marks; here it only aborts (the SDK silences the late one).
            return CallOutcome::Err {
                code: "cancelled".to_string(),
                message: "aborted".to_string(),
            };
        }
        if let Some(code) = input.get("fail").and_then(|v| v.as_str()) {
            self.marks.lock().unwrap().remove(ticket);
            return CallOutcome::Err {
                code: code.to_string(),
                message: format!("remote {}", code),
            };
        }
        if let Some(fl) = input.get("flood_stream") {
            let fid = fl.get("id").and_then(|v| v.as_str()).unwrap_or("s1");
            let chunk = fl.get("chunk").and_then(|v| v.as_u64()).unwrap_or(65536).min(1 << 20) as usize;
            let count = fl.get("count").and_then(|v| v.as_u64()).unwrap_or(4).min(64);
            let payload = "x".repeat(chunk);
            for seq in 0..count {
                if cancel.load(Ordering::SeqCst) {
                    return CallOutcome::Err {
                        code: "cancelled".to_string(),
                        message: "aborted".to_string(),
                    };
                }
                if ctx.send_stream(fid, seq, &payload).is_err() {
                    break;
                }
            }
        }
        self.marks.lock().unwrap().remove(ticket);
        CallOutcome::Ok(json_echo(input))
    }

    fn on_cancel(&self, ticket: &str) {
        if let Some(path) = self.marks.lock().unwrap().remove(ticket) {
            append_line(&path, ticket);
        }
    }
}

fn json_echo(input: &Value) -> Value {
    serde_json::json!({"echo": input})
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut sock = String::new();
    let mut id = "ext-echo".to_string();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--matrix-sock" => {
                i += 1;
                sock = args.get(i).cloned().unwrap_or_default();
            }
            "--id" => {
                i += 1;
                id = args.get(i).cloned().unwrap_or(id);
            }
            _ => {}
        }
        i += 1;
    }
    if sock.is_empty() {
        eprintln!("usage: ext_echo --matrix-sock <sock> [--id <logical>]");
        std::process::exit(2);
    }
    let comp = Component::connect(std::path::Path::new(&sock), &id).expect("connect");
    match comp.serve(Echo { marks: Mutex::new(HashMap::new()) }) {
        Ok(()) => std::process::exit(0),
        Err(e) if e.contains("eof") => std::process::exit(0),
        Err(e) => {
            eprintln!("serve: {}", e);
            std::process::exit(1);
        }
    }
}
